src/main/java/org/jabref/gui/mergeentries/newmergedialog/FieldRowController.java

```diff
@@ -0,0 +1,139 @@
+package org.jabref.gui.mergeentries.newmergedialog;
+
+import javafx.beans.property.ReadOnlyStringProperty;
+import javafx.scene.control.ToggleGroup;
+
+import org.jabref.gui.mergeentries.newmergedialog.cell.FieldNameCell;
+import org.jabref.gui.mergeentries.newmergedialog.cell.FieldValueCell;
+import org.jabref.gui.mergeentries.newmergedialog.cell.MergedFieldCell;
+import org.jabref.gui.mergeentries.newmergedialog.diffhighlighter.SplitDiffHighlighter;
+import org.jabref.gui.mergeentries.newmergedialog.diffhighlighter.UnifiedDiffHighlighter;
+import org.jabref.gui.mergeentries.newmergedialog.toolbar.ThreeWayMergeToolbar;
+import org.jabref.model.strings.StringUtil;
+
+import org.fxmisc.richtext.StyleClassedTextArea;
+
+public class FieldRowController {
```

Comment:
A controller? Doesn't this conflict with the mvvm pattern we are using?

Reply:
The idea was that a field row would take model data (left and right field values), create the views and bind the model to it. A controller would do exactly that. It cannot be a view-model because it contains a reference to the view, and it cannot be a view because it doesn't inherit the Node class.

Also, [JabRef architecture docs](https://devdocs.jabref.org/readings-on-coding/javafx.html) state that because FXML is not powerful enough and it can't do all the binding, we need a controller. So a controller does have a place in our architecture. The only new thing that I'm introducing here is the `controller` keyword, which I plan to remove and I removed while implementing the [merging groups feature](https://github.com/JabRef/jabref/pull/8961/files#diff-6144c7471727235c86c556836c1a1d3d28586e91b8ad75100348c35088feb48e).

Other than that, I'll appreciate any ideas or hints you have about making the current design better.

Reply:
I was just wondering about the word.
Yes sometimes we do more in the java view class beyond binding nodes from the parsed fxml to the viewmodel (see https://www.ctan.org/tex-archive/biblio/bibtex/contrib/doc/ ), because fxml is not powerful enough and there are architectural limitations in java/javafx. However, I still see some business logic in the controller, that can be pushed into a viewmodel to be able to test it, like keeping the values of the cells and `hasEqualLeftAndRightValues`. It would just be great to be able to test the logic of your custom row.
Btw, _not using_ FXML is totally ok with me, because parsing FXML takes time - a lot of time. So I think you already did it right not to introduce an FXML for the row, since parsing 10+ fxml files is wasting too much cpu time.

Reply:
I agree with Carl, rename this, just call it FieldRow
and create a new class FieldRowViewModel where you put the logic.
 You would bind it then e.g fieldNameCell.text().bindBidirectional(viewModel.fieldNameText)